src/main/java/ca/dmoj/java/SubmissionAgent.java

```diff
@@ -22,29 +21,27 @@ public static void premain(String argv, Instrumentation inst) throws Unsupported
 
         final Thread selfThread = Thread.currentThread();
 
+        if (!unsafe) {
+            inst.addTransformer(new DisallowedClassesClassFileTransformer(
+                new String[]{
+                    "sun/misc/Unsafe",
+                },
+                new RuntimeException[]{
+                    new UnsafeDisallowedException()
+            }));
+        }
         if (noBigMath) {
-            inst.addTransformer(new ClassFileTransformer() {
-                @Override
-                public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined,
-                                        ProtectionDomain protectionDomain, byte[] classfileBuffer)
-                                        throws IllegalClassFormatException {
-                    if (className == null) return classfileBuffer;
-
-                    RuntimeException disallowed = null;
-                    // If the class ever loaded it's because a submission used it
-                    if (className.startsWith("java/math/BigInteger") ||
-                        className.startsWith("java/math/MutableBigInteger")) {
-                        disallowed = new BigIntegerDisallowedException();
-                    } else if (className.startsWith("java/math/BigDecimal")) {
-                        disallowed = new BigDecimalDisallowedException();
-                    }
-
-                    if (disallowed != null) dumpExceptionAndExit(disallowed);
-
-                    // Don't actually retransform anything
-                    return classfileBuffer;
-                }
-            });
+            inst.addTransformer(new DisallowedClassesClassFileTransformer(
+                new String[]{
+                    "java/math/BigInteger",
+                    "java/math/MutableBigInteger",
+                    "java/math/BigDecimal"
+                },
+                new RuntimeException[]{
+                    new BigIntegerDisallowedException(),
+                    new BigIntegerDisallowedException(),
+                    new BigDecimalDisallowedException(),
+            }));
```

Comment:
I don't like the idea of using two arrays that magically needs to be the same length...

Reply:
I haven't coded in Java in a while, what do you suggest I use instead? Hashmap?

Reply:
You could use sort sort of a map, but that seems overkill. Just generate a class for `DisallowedClassRule` that takes the two parameters in the constructor is probably enough. Also pre-creating the exception feels a bit yucky so maybe you can take an interface to receive `BigIntegerDisallowedException::new` et al., but whatever.